container/src/main/java/com/flipkart/poseidon/filters/DebugFilter.java

```diff
@@ -0,0 +1,277 @@
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpServletResponseWrapper;
+import java.io.*;
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * Created by shrey.garg on 30/05/17.
+ */
+public class DebugFilter implements Filter {
+    @Override
+    public void init(FilterConfig filterConfig) throws ServletException {
+
+    }
+
+    @Override
+    public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse, FilterChain filterChain) throws IOException, ServletException {
+        HttpServletRequest httpServletRequest;
+        ServletResponse response = servletResponse;
+
+        if (servletRequest instanceof HttpServletRequest) {
+            httpServletRequest = (HttpServletRequest) servletRequest;
+
+            if (httpServletRequest.getParameter("whatIsWrongWithThis") != null) {
+                ServiceContext.enableDebug();
+                response = new HttpServletResponseCopier((HttpServletResponse) servletResponse);
+            }
+
+            filterChain.doFilter(httpServletRequest, response);
+        } else {
+            filterChain.doFilter(servletRequest, servletResponse);
+        }
+
+        if (ServiceContext.isDebug()) {
+            httpServletRequest = (HttpServletRequest) servletRequest;
+
+            String includeOnlyServices = httpServletRequest.getParameter("includeOnly");
+            boolean disableServiceResponses = httpServletRequest.getParameterMap().containsKey("disableServiceResponses");
+
+            Map<String, List<ServiceDebug>> callDebug = filterWantedServices(includeOnlyServices, ServiceContext.getDebugResponses());
+            filterDisabledServices(httpServletRequest, callDebug);
+
+            Map<String, List<ServiceCallDebug>> serviceCallDebugMap = new HashMap<>();
+            for (Map.Entry<String, List<ServiceDebug>> entry : callDebug.entrySet()) {
+                List<ServiceDebug> serviceDebugs = entry.getValue();
+                if (serviceDebugs == null) {
+                    continue;
+                }
+
+                List<ServiceCallDebug> serviceCallDebugs = new ArrayList<>();
+                for (ServiceDebug serviceDebug : serviceDebugs) {
+                    ServiceCallDebug serviceCallDebug = convertToServiceCallDebug(serviceDebug, disableServiceResponses);
+                    serviceCallDebugs.add(serviceCallDebug);
+                }
+
+                serviceCallDebugMap.put(entry.getKey(), serviceCallDebugs);
+            }
+
+            generateDebugResponse((HttpServletResponseCopier) response, serviceCallDebugMap);
+        }
+    }
+
+    /**
+     * The services corresponding values contained in this map will be not be sent as part
+     * of the debug response unless the key is present as a query parameter in the request.
+     * @return a mapping of enable flags to service client class names
+     */
+    protected Map<String, String> enableFlags() {
+        return new HashMap<>();
```

Comment:
This is little confusing. Query Parameter to look for should be the key and the service client class to skip is the value. Instead of we binding it a query parameter, can we change this syntax to
`List<String>  getServiceClientsToDisable(HttpServletRequest request)`

Sub classes can return a list of service client class names that has to be removed and this decision can be dynamically made using any parameter from request.